frontend/web/src/app/components/material/material-edit/material-edit.component.ts

```diff
@@ -0,0 +1,256 @@
+    const fields = {
+      name: [this.material && this.material.name ? this.material.name : '',
+        Validators.compose([Validators.required])],
+      supplierName: [this.material && this.material.supplierName ? this.material.supplierName : '',
+        Validators.compose([Validators.required])],
+      initialAmount: [this.material && this.material.initialAmount ? this.material.initialAmount : '',
+        Validators.compose([Validators.required])],
+    };
+
+    this.requiredFieldsForm = this.formBuilder.group(fields);
+    this.loaded = true;
+  }
+
+  private constructMaterialGradeForms() {
+    /* load attr with default data from product instance*/
+    this.materialGradeFieldsForm = [];
+
+    this.material.materialGrades.forEach((materialGrade) => {
+      this.materialGradeFieldsForm.push(this.formBuilder.group({
+        reusedTimes: [materialGrade.reusedTimes, Validators.compose([Validators.required])],
+        amount: [materialGrade.amount, Validators.compose([Validators.required])],
+      }));
+    });
+  }
+
+  private constructPropertiesForms() {
+    /* load attr with default data from product instance*/
+    this.materialPropertiesFieldsForm = [];
+    Object.getOwnPropertyNames(this.material.materialProperties).forEach((id) => {
+      this.materialPropertiesFieldsForm.push(this.formBuilder.group({
+        value: [id,
+          Validators.compose([Validators.required])],
+        property: [this.material.materialProperties[id] ? this.material.materialProperties[id] : '',
+          Validators.compose([Validators.required])],
+      }));
+    });
+  }
+
+  /* save product instance */
+  save() {
+    /* convert relevant fields */
+
+    const id = this.material.id;
+    const material: Material = new Material(this.requiredFieldsForm.value);
+
+    this.creating ? delete material['id'] : material.id = id;
+    const materialGrades: MaterialGrade[] = new Array<MaterialGrade>();
+    const materialProperties = {};
+    this.materialGradeFieldsForm.forEach((formGroup) => {
+      materialGrades.push(new MaterialGrade({
+        amount: formGroup.value.amount,
+        reusedTimes: formGroup.value.reusedTimes,
+      }));
+    });
+    material.materialGrades = materialGrades;
+    this.materialPropertiesFieldsForm.forEach((formGroup) => {
+      const v = formGroup.value.value;
+      const p = formGroup.value.property;
+      materialProperties[v] = p;
+    });
+    material.materialProperties = materialProperties;
+    if (this.creating) { // a new product
+      this.materialService.createMaterial(material).subscribe(
+        (data) => {
+          if (this.nav) {
+            this.back();
+          }
+          this.changed.emit(data);
+        }, (error) => {
+          console.log(error);
+          this.errorService.showAlert(error.verobose_message_header, error.verbose_message);
+        },
+      );
+    } else {
+      this.materialService.updateMaterial(material).subscribe(
+        (data) => {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Update this function so that its implementation is not identical to the one on line 157. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=typescript%3AS4144)